pkg/util/webconsole/proxy_address.go

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2024 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package webconsole
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	appv1a1 "github.com/vmware-tanzu/vm-operator/external/appplatform/api/v1alpha1"
+	pkgcfg "github.com/vmware-tanzu/vm-operator/pkg/config"
+)
+
+const (
+	SupervisorServiceObjName      = "supervisor-env-props"
+	SupervisorServiceObjNamespace = "vmware-system-supervisor-services"
+
+	ProxyAddrServiceName      = "kube-apiserver-lb-svc"
+	ProxyAddrServiceNamespace = "kube-system"
+)
+
+// ProxyAddress first attempts to get the proxy address through the API Server DNS Names.
+// If that is unset, though, fall back to using the virtual IP.
+func ProxyAddress(ctx context.Context, r client.Client) (string, error) {
+	if !pkgcfg.FromContext(ctx).Features.SimplifiedEnablement {
+		return proxyAddressFromVirtualIP(ctx, r)
+	}
+
+	// Attempt to use the API Server DNS Names to get the proxy address.
+	proxyAddress, err := proxyServiceDNSName(ctx, r)
+	if err != nil {
+		return "", fmt.Errorf("failed to get proxy service URL: %w", err)
+	}
+
+	// If no API Server DNS Name exists, fall back to using the virtual IP.
+	if proxyAddress == "" {
+		return proxyAddressFromVirtualIP(ctx, r)
+	}
+
+	return proxyAddress, nil
+}
+
+// proxyServiceDNSName retrieves the first API server DNS name using the provided client by
+// querying the appplatform CRD, if one exists.
+func proxyServiceDNSName(ctx context.Context, r client.Client) (string, error) {
+	proxySvc := &appv1a1.SupervisorProperties{}
+	proxySvcObjectKey := client.ObjectKey{Name: SupervisorServiceObjName, Namespace: SupervisorServiceObjNamespace}
+
+	if err := r.Get(ctx, proxySvcObjectKey, proxySvc); err != nil {
+		return "", fmt.Errorf("failed to get proxy address service %s: %w", proxySvcObjectKey, err)
```

Comment:
Ignore the "not found" error here to fall back to virtual IP when the CR doesn't exist?

Reply:
Done, also updated the comment.

Reply:
Hey sorry if I wasn't clear in my previous comment. What I meant was to only ignore the "not-found" error since there is no point retrying if the CR doesn't exist in cluster, something like:

```suggestion
import (
    ...
    apierrors "k8s.io/apimachinery/pkg/api/errors"
)

func proxyServiceDNSName(ctx context.Context, r client.Client) (string, error) {
	proxySvc := &appv1a1.SupervisorProperties{}
	proxySvcObjectKey := client.ObjectKey{Name: SupervisorServiceObjName, Namespace: SupervisorServiceObjNamespace}

	if err := r.Get(ctx, proxySvcObjectKey, proxySvc); err != nil {
		if apierrors.IsNotFound(err) {
			return "", nil
		}
		return "", fmt.Errorf("failed to get proxy address service %s: %w", proxySvcObjectKey, err)
	}
	...
```

Reply:
Fixed.